### PR TITLE
Adds a Shift expression

### PR DESCRIPTION
**Pull request by cpcloud**

### CODE REVIEW

File: blaze/expr/collections.py [hunk 1]
```diff
@@ -708,8 +710,45 @@ def isin(expr, keys):
...
+    """
+    __slots__ = '_hash', '_child', 'n'
+
+    @property
+    def schema(self):
+        measure = self._child.schema.measure
+
+        # if we are not shifting or we are already an Option type then return
+        # the child's schema
+        if not self.n or isinstance(measure, Option):
```
Comment (llllllllll) [inline blaze/expr/collections.py]: should a shift(expr, 0) just return expr?

File: blaze/expr/collections.py [hunk 2]
```diff
@@ -708,8 +710,45 @@ def isin(expr, keys):
...
+
+    def __str__(self):
+        return '%s.%s(n=%d)' % (
+            self._child, type(self).__name__.lower(), self.n
+        )
+
+
+@copydoc(Shift)
+def shift(expr, n):
+    if not isinstance(n, (numbers.Integral, np.integer)):
```
Comment (llllllllll) [inline blaze/expr/collections.py]: is there a reason we cannot allow expressions for n. like, `a.shift(b.count())`?

File: blaze/expr/collections.py [hunk 3]
```diff
@@ -708,8 +710,49 @@ def isin(expr, keys):
     return IsIn(expr, frozenset(keys))
 
 
+class Shift(Expr):
+    """ Shift a column backward or forward by N elements
+
+    Parameters
+    ----------
+    expr : Expr
+        The expression to shift. This expression's dshape should be columnar
```
Comment (llllllllll) [inline blaze/expr/collections.py]: Should we typecheck in the `shift` constructor?

Comment (cpcloud) [inline blaze/expr/collections.py]: no, this done by the dshape methods/property checking in `__getattr__`

File: blaze/expr/collections.py (hunk 2 above)
Comment (cpcloud) [inline blaze/expr/collections.py]: No, but I'd like to do that in a separate PR

File: blaze/expr/collections.py (hunk 3 above)
Comment (llllllllll) [inline blaze/expr/collections.py]: Should we not expose the `shift` toplevel function also?

File: blaze/expr/collections.py (hunk 2 above)
Comment (llllllllll) [inline blaze/expr/collections.py]: kk

File: blaze/expr/collections.py (hunk 1 above)
Comment (cpcloud) [inline blaze/expr/collections.py]: I'm going to leave it as is for now. If this becomes a bottleneck in the future then it's very easy to change
